### ISSUE TITLE
Sanity check on 'url', 'hostname' and 'domain' indicators

### ISSUE THREAD

**Issue by kjellchr** [MISP/MISP#3506]
The content of this ticket is describing issues where the 'IDS' flag is set to true.
Is it possible to get sanity checks to avoid the examples from below from happening? If any of these are expected behaviour please let me know so we can adjust our tools accordingly.

**URL**

Every now and then errors in the 'url' field is observed, e.g. there might be a space in it, not respresented by %20 but an actual space in it. Other times it one would observe '[' and ']' in the domain name. The efforts generally appear to be lack of defanging URLs, but sometimes it also looks like simply misunderstanding what should be put into the field, e.g. '$domain - GET' to tell the recipient that what one should look for is GET requests. Of course this doesn't help automated mechanisms that is written to compare signatures.

Other times, there are simply regex or substrings in the 'url' field - none of, to my understanding, the field was made for (happy to take a link to documentation that says otherwise, if it exists - I could not find any).

*examples:*
```
\/example_string$
\/example\/(more|of|the|strings)$
example/test.php
http://examp le.com
http://example .com/badurl
http://alpha. example.com/
\\1.2.3.4\example\file.exe
.example&alpha=1&bravo=randomstring
Example=
example=alpha
http://example..com/alpha
http :// example .com / Alpha / bravo.php
www.example.com - GET
htpp://example.com/alpha
alpha@example.com
1.2.3.4example
example.com.
http:/example.com/alpha
1.2.3
http://example,com
http://example.com /
http://example.com./
http://example.com:1234:1234/index.php
http://example.com:80:/alpha
http://&#8203;example.com/alpha/
http&#8203;://example.com/alpha
h&#8203;ttp://example.com/alpha
http:\\example.com/alpha
http://1.2.3.4example/alpha.exe
example.com|1.2.3.4
http://example/alpha
ttps://example.com
https://example[.com
hhttp://example.com/alpha
http://.*/example
https://*.example.com/*
http:]//example.com/alpha
https:///example.com
```

**Hostname**

hostnames with multiple '.' after each other have been observed. Should this be possible?

*examples:*
```
alpha..example.com
bravo..example.com
```

**Domain**

In some cases prepended '.' is observed. is it defined anywhere that this means 'all subdomains of a domain', or some such? I *think* that is what is made (that would make sense) but I am unable to find documentation that confirms this.

*example:*
```
.example.com
```

**Comment by iglocska** [comment]
Yeah, this comes up quite often. I don't think I'd be in favour of doing the sanity check on ingestion side for URLs, though the hostname + domain examples make sense.

On the other hand we should build the tighter controls into the IDS exports specifically.

**Comment by kjellchr** [comment]
Any reason why there should not be a sanity check on the ingestion side (assuming the IDS flag is checked)? More curious on the thought process around this.

In a lot of cases the user would (probably) realise the issue and fix it if he/she gets some sort of warning that this is in fact not a valid URL.

The biggest part of this issue would likely be resolved if the IDS exports had sanity checks, however the issue then becomes than you might drop a bunch of URLs you would want to match on, however the time/moment where this should had been fixed has long since passed.

**Comment by iglocska** [comment]
In some cases you want to capture malformed information in malicious samples for example as they're also a valid artifact. We do the same for example with malformed financial indicators, but there we show a warning of it being malformed - maybe we need the same for urls. On the other hand, different tools deal differently with malformed URLs. Some will allow partial URLs as valid input, some will not - this should be up to the export system to filter them out.

**Comment by kjellchr** [comment]
Showing a warning would makes sense. They are indeed a valid artifact, depending on the context.

More often than not, IDS type tools will not be able to deal with malformed info as the whatever would generate them would most likely not even be able to generate network traffic.

By showing a warning (perhaps an option letting people choose 'I know this is malformed, please go ahead and store it') and then marking it as malformed and not exporting what is malformed (or at least giving that as an option), I believe we would resolve the issue - prevent garbage from being imported into tools while at the same time giving proper warning for folks fat-fingering indicators.
